compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs

```diff
@@ -464,7 +464,26 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
                                 root_obligation,
                             )
                         } else {
-                            (trait_predicate, &obligation)
+                            let mut temp_trait_predicate = trait_predicate;
+                            let mut loop_code = obligation.cause.code().clone();
+                            loop{
+                                let (code, base) = loop_code.peel_derives_with_predicate();
+                                if matches!(loop_code, ObligationCauseCode::ImplDerivedObligation(ref c) if tcx.has_attr(c.impl_or_alias_def_id, sym::do_not_recommend)) {
+                                    if let Some(base) = base {
+                                        let code = code.clone();
+                                        obligation.cause.map_code(|_| code);
+                                        obligation.predicate = ty::PredicateKind::Clause(ty::ClauseKind::Trait(base.skip_binder().clone())).to_predicate(tcx);
```

Comment:
Let's not skip the binder here -- it will almost certainly cause an ICE later on. [`PolyTraitPredicate<'tcx>: ToPredicate<'tcx>`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/trait.ToPredicate.html#impl-ToPredicate%3C'tcx%3E-for-Binder%3C'tcx,+TraitPredicate%3C'tcx%3E%3E) should do the trick.

```suggestion
                                        obligation.predicate = base.to_predicate(tcx);
```